File: internal/runbook/runbook.go
package runbook

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/newrelic/nr-diagnose/internal/schemas"
	"gopkg.in/yaml.v3"
)

// Manager handles runbook loading, matching, and writing.
type Manager struct {
	seededDir string
	localDir  string
	seeded    schemas.RunbookIndex
	local     schemas.RunbookIndex
}

// NewManager creates a runbook manager from seeded and local directories.
func NewManager(seededDir, localDir string) (*Manager, error) {
	m := &Manager{
		seededDir: seededDir,
		localDir:  localDir,
	}

	if seededDir != "" {
		m.seeded = loadIndex(filepath.Join(seededDir, "index.yaml"))
	}

	if localDir != "" {
		m.local = loadIndex(filepath.Join(localDir, "index.yaml"))
	}

	return m, nil
}

// Match checks both seeded and local runbooks for a matching error pattern.
func (m *Manager) Match(errorOutput string) (schemas.RunbookEntry, bool) {
	lowerError := strings.ToLower(errorOutput)

	if entry, found := m.matchIndex(m.seeded, m.seededDir, lowerError); found {
		return entry, true
	}

	if entry, found := m.matchIndex(m.local, m.localDir, lowerError); found {
		return entry, true
	}

	return schemas.RunbookEntry{}, false
}

// WriteEntry saves a new runbook entry to the local directory.
func (m *Manager) WriteEntry(agentName string, entry schemas.RunbookEntry) error {
	if m.localDir == "" {
		return fmt.Errorf("no local runbook directory configured")
	}

	agentRunbook := filepath.Join(m.localDir, agentName)
	os.MkdirAll(agentRunbook, 0755)

	entry.ID = fmt.Sprintf("%03d", len(m.local.Entries)+1)
	entry.FirstSeen = time.Now().UTC().Format(time.RFC3339)
	entry.LastSeen = entry.FirstSeen
	entry.ResolvedCount = 1

	slug := slugify(entry.RootCause)
	filename := fmt.Sprintf("%s-%s.md", entry.ID, slug)
	entryPath := filepath.Join(agentRunbook, filename)

	content := fmt.Sprintf("---\nid: \"%s\"\nerror_pattern: \"%s\"\nstep_failed: \"%s\"\nroot_cause: \"%s\"\nfix_command: \"%s\"\nresolved_count: %d\nfirst_seen: \"%s\"\nlast_seen: \"%s\"\n---\n\n## Resolution\n`%s`\n", entry.ID, entry.ErrorPattern, entry.StepFailed, entry.RootCause, entry.FixCommand, entry.ResolvedCount, entry.FirstSeen, entry.LastSeen, entry.FixCommand)

	if err := os.WriteFile(entryPath, []byte(content), 0644); err != nil {
		return err
	}

	m.local.Entries = append(m.local.Entries, schemas.RunbookIndexEntry{
		Pattern:   entry.ErrorPattern,
		EntryFile: filename,
	})

	return m.saveLocalIndex(agentRunbook)
}

// IncrementCount updates the resolved_count and last_seen for an existing entry.
func (m *Manager) IncrementCount(entry *schemas.RunbookEntry) {
	entry.ResolvedCount++
	entry.LastSeen = time.Now().UTC().Format(time.RFC3339)
}

func (m *Manager) matchIndex(index schemas.RunbookIndex, dir string, lowerError string) (schemas.RunbookEntry, bool) {
	for _, ie := range index.Entries {
		pattern := strings.ToLower(ie.Pattern)
		if strings.Contains(lowerError, pattern) {
			entry, err := loadEntry(filepath.Join(dir, ie.EntryFile))
			if err == nil {
				return entry, true
			}
		}
	}
	return schemas.RunbookEntry{}, false
}

func (m *Manager) saveLocalIndex(dir string) error {
	data, err := yaml.Marshal(m.local)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "index.yaml"), data, 0644)
}

func loadIndex(path string) schemas.RunbookIndex {
	data, err := os.ReadFile(path)
	if err != nil {
		return schemas.RunbookIndex{}
	}
	var index schemas.RunbookIndex
	yaml.Unmarshal(data, &index)
	return index
}

func loadEntry(path string) (schemas.RunbookEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return schemas.RunbookEntry{}, err
	}

	content := string(data)
	parts := strings.SplitN(content, "---", 3)
	if len(parts) < 3 {
		return schemas.RunbookEntry{}, fmt.Errorf("invalid frontmatter in %s", path)
	}

	var entry schemas.RunbookEntry
	if err := yaml.Unmarshal([]byte(parts[1]), &entry); err != nil {
		return schemas.RunbookEntry{}, err
	}
	return entry, nil
}

func slugify(s string) string {
	s = strings.ToLower(s)
	s = strings.Map(func(r rune) rune {
		if r >= 'a' && r <= 'z' || r >= '0' && r <= '9' {
			return r
		}
		return '-'
	}, s)
	for strings.Contains(s, "--") {
		s = strings.ReplaceAll(s, "--", "-")
	}
	s = strings.Trim(s, "-")
	if len(s) > 40 {
		s = s[:40]
	}
	return s
}

File: nr_diagnose/runbook.py
"""Runbook manager - handles loading, matching, and writing runbook entries."""

import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional, Tuple

import yaml

from .schemas import RunbookEntry, RunbookIndex, RunbookIndexEntry


class Manager:
    """Handles runbook loading, matching, and writing."""

    def __init__(self, seeded_dir: str, local_dir: str):
        self.seeded_dir = seeded_dir
        self.local_dir = local_dir
        self.seeded = _load_index(os.path.join(seeded_dir, "index.yaml")) if seeded_dir else RunbookIndex()
        self.local = _load_index(os.path.join(local_dir, "index.yaml")) if local_dir else RunbookIndex()

    def match(self, error_output: str) -> Tuple[Optional[RunbookEntry], bool]:
        """Check both seeded and local runbooks for a matching error pattern.

        Returns (entry, found).
        """
        lower_error = error_output.lower()

        entry = self._match_index(self.seeded, self.seeded_dir, lower_error)
        if entry:
            return entry, True

        entry = self._match_index(self.local, self.local_dir, lower_error)
        if entry:
            return entry, True

        return None, False

    def write_entry(self, agent_name: str, entry: RunbookEntry) -> None:
        """Save a new runbook entry to the local directory."""
        if not self.local_dir:
            raise RuntimeError("no local runbook directory configured")

        # cli.py already constructs local_dir as <data>/runbook/<agent_name>, so don't
        # nest another <agent_name> level here. agent_name is still used to derive the
        # index path passed to _save_local_index below.
        agent_runbook = self.local_dir
        os.makedirs(agent_runbook, exist_ok=True)

        entry.id = f"{len(self.local.entries) + 1:03d}"
        entry.first_seen = datetime.now(timezone.utc).isoformat()
        entry.last_seen = entry.first_seen
        entry.resolved_count = 1

        slug = _slugify(entry.root_cause)
        filename = f"{entry.id}-{slug}.md"
        entry_path = os.path.join(agent_runbook, filename)

        content = f"""---
id: "{entry.id}"
error_pattern: "{entry.error_pattern}"
step_failed: "{entry.step_failed}"
root_cause: "{entry.root_cause}"
fix_command: "{entry.fix_command}"
resolved_count: {entry.resolved_count}
first_seen: "{entry.first_seen}"
last_seen: "{entry.last_seen}"
---

## Resolution
`{entry.fix_command}`
"""
        Path(entry_path).write_text(content)

        self.local.entries.append(RunbookIndexEntry(
            pattern=entry.error_pattern,
            entry_file=filename,
        ))

        self._save_local_index(agent_runbook)

    def increment_count(self, entry: RunbookEntry) -> None:
        """Update resolved_count and last_seen for an existing entry."""
        entry.resolved_count += 1
        entry.last_seen = datetime.now(timezone.utc).isoformat()

    def _match_index(self, index: RunbookIndex, dir_path: str, lower_error: str) -> Optional[RunbookEntry]:
        for ie in index.entries:
            pattern = ie.pattern.lower()
            if pattern in lower_error:
                entry = _load_entry(os.path.join(dir_path, ie.entry_file))
                if entry:
                    return entry
        return None

    def _save_local_index(self, dir_path: str) -> None:
        data = {
            "entries": [
                {"pattern": e.pattern, "entry_file": e.entry_file}
                for e in self.local.entries
            ]
        }
        Path(os.path.join(dir_path, "index.yaml")).write_text(yaml.dump(data))


def _load_index(path: str) -> RunbookIndex:
    """Load a runbook index from a YAML file."""
    try:
        data = yaml.safe_load(Path(path).read_text()) or {}
        entries = [
            RunbookIndexEntry(pattern=e.get("pattern", ""), entry_file=e.get("entry_file", ""))
            for e in data.get("entries", [])
        ]
        return RunbookIndex(entries=entries)
    except (FileNotFoundError, OSError):
        return RunbookIndex()


def _load_entry(path: str) -> Optional[RunbookEntry]:
    """Load a runbook entry from a frontmatter markdown file."""
    try:
        content = Path(path).read_text()
        parts = content.split("---", 2)
        if len(parts) < 3:
            return None

        data = yaml.safe_load(parts[1])
        if not data:
            return None

        return RunbookEntry(
            id=data.get("id", ""),
            error_pattern=data.get("error_pattern", ""),
            step_failed=data.get("step_failed", ""),
            root_cause=data.get("root_cause", ""),
            fix_command=data.get("fix_command", ""),
            resolved_count=data.get("resolved_count", 0),
            first_seen=data.get("first_seen", ""),
            last_seen=data.get("last_seen", ""),
        )
    except (FileNotFoundError, OSError):
        return None


def _slugify(s: str) -> str:
    """Convert a string to a URL-safe slug."""
    s = s.lower()
    s = re.sub(r'[^a-z0-9]', '-', s)
    s = re.sub(r'-+', '-', s)
    s = s.strip('-')
    return s[:40]
